### ISSUE TITLE
"Can not find chunk parent during dependency sort" issue with webpack2

### ISSUE THREAD

**Issue by michael-wolfenden** [ampedandwired/html-webpack-plugin#329]
## Issue

With the following webpack.config.js

```javascript
var path = require('path');
var webpack = require('webpack');
var HtmlWebpackPlugin = require('html-webpack-plugin')

module.exports = {
    entry: './index.js',

    output: {
        path: path.join(__dirname, 'build'),
        filename: '[name].[chunkhash].js',
        chunkFilename: '[chunkhash].js'
    },

    plugins: [
        new webpack.optimize.CommonsChunkPlugin({
            names: ['manifest']
        }),

        new HtmlWebpackPlugin({
            excludeChunks: ['manifest']
        })
    ]
};
```

Running webpack2 gives the following error

```
Unhandled rejection Error: Can not find chunk parent during dependency sort
    at D:\Dev\html-webpack-plugin-manifest-issue\node_modules\html-webpack-plugin\lib\chunksorter.js:45:17
    at Array.forEach (native)
    at D:\Dev\html-webpack-plugin-manifest-issue\node_modules\html-webpack-plugin\lib\chunksorter.js:42:21
    at Array.forEach (native)
    at Object.module.exports.dependency (D:\Dev\html-webpack-plugin-manifest-issue\node_modules\html-webpack-plugin\lib\chunksorter.js:39:10)
    at HtmlWebpackPlugin.sortChunks (D:\Dev\html-webpack-plugin-manifest-issue\node_modules\html-webpack-plugin\index.js:329:33)
    at Compiler.<anonymous> (D:\Dev\html-webpack-plugin-manifest-issue\node_modules\html-webpack-plugin\index.js:70:19)
    at Compiler.applyPluginsAsync (D:\Dev\html-webpack-plugin-manifest-issue\node_modules\tapable\lib\Tapable.js:85:13)
    at Compiler.emitAssets (D:\Dev\html-webpack-plugin-manifest-issue\node_modules\webpack\lib\Compiler.js:260:7)
    at Compiler.onCompiled (D:\Dev\html-webpack-plugin-manifest-issue\node_modules\webpack\lib\Compiler.js:202:11)
    at D:\Dev\html-webpack-plugin-manifest-issue\node_modules\webpack\lib\Compiler.js:437:12
    at next (D:\Dev\html-webpack-plugin-manifest-issue\node_modules\tapable\lib\Tapable.js:81:11)
    at Compiler.<anonymous> (D:\Dev\html-webpack-plugin-manifest-issue\node_modules\webpack\lib\CachePlugin.js:61:4)
    at Compiler.applyPluginsAsync (D:\Dev\html-webpack-plugin-manifest-issue\node_modules\tapable\lib\Tapable.js:85:13)
    at Compiler.<anonymous> (D:\Dev\html-webpack-plugin-manifest-issue\node_modules\webpack\lib\Compiler.js:434:9)
    at Compilation.<anonymous> (D:\Dev\html-webpack-plugin-manifest-issue\node_modules\webpack\lib\Compilation.js:577:13)
    at Compilation.applyPluginsAsync (D:\Dev\html-webpack-plugin-manifest-issue\node_modules\tapable\lib\Tapable.js:73:70)
    at Compilation.<anonymous> (D:\Dev\html-webpack-plugin-manifest-issue\node_modules\webpack\lib\Compilation.js:572:10)
    at Compilation.applyPluginsAsync (D:\Dev\html-webpack-plugin-manifest-issue\node_modules\tapable\lib\Tapable.js:73:70)
    at Compilation.<anonymous> (D:\Dev\html-webpack-plugin-manifest-issue\node_modules\webpack\lib\Compilation.js:567:9)
    at Compilation.applyPluginsAsync (D:\Dev\html-webpack-plugin-manifest-issue\node_modules\tapable\lib\Tapable.js:73:70)
    at Compilation.<anonymous> (D:\Dev\html-webpack-plugin-manifest-issue\node_modules\webpack\lib\Compilation.js:563:8)
    at Compilation.applyPluginsAsync (D:\Dev\html-webpack-plugin-manifest-issue\node_modules\tapable\lib\Tapable.js:73:70)
    at Compilation.seal (D:\Dev\html-webpack-plugin-manifest-issue\node_modules\webpack\lib\Compilation.js:521:7)
    at Compiler.<anonymous> (D:\Dev\html-webpack-plugin-manifest-issue\node_modules\webpack\lib\Compiler.js:431:15)
    at D:\Dev\html-webpack-plugin-manifest-issue\node_modules\tapable\lib\Tapable.js:152:11
    at D:\Dev\html-webpack-plugin-manifest-issue\node_modules\html-webpack-plugin\index.js:60:9
    at tryCatcher (D:\Dev\html-webpack-plugin-manifest-issue\node_modules\bluebird\js\release\util.js:16:23)
    at Promise._settlePromiseFromHandler (D:\Dev\html-webpack-plugin-manifest-issue\node_modules\bluebird\js\release\promise.js:502:31)
    at Promise._settlePromise (D:\Dev\html-webpack-plugin-manifest-issue\node_modules\bluebird\js\release\promise.js:559:18)
```

## Reproducing

I have a repository reproducing the issue at:
https://github.com/michael-wolfenden/html-webpack-plugin-manifest-issue

## Investigation

It looks like for webpack2, the default dependency sort changed from id to dependency

```
// In webpack 2 the ids have been flipped.
// Therefore the id sort doesn't work the same way as it did for webpack 1
// Luckily the dependency sort is working as expected
if (require('webpack/package.json').version.split('.')[0] === '2') {
  module.exports.auto = module.exports.dependency;
}
```

The issue is that if you look at the chunks

```json
[
    {
        "id": 1,
        "rendered": true,
        "initial": true,
        "entry": false,
        "extraAsync": false,
        "size": 57,
        "names": [
            "main"
        ],
        "files": [
            "main.fdb3f583baeb7d160bcc.js"
        ],
        "hash": "fdb3f583baeb7d160bcc",
        "parents": [
            0
        ],
        "origins": [
            {
                "moduleId": 0,
                "module": "D:\\Dev\\html-webpack-plugin-manifest-issue\\index.js",
                "moduleIdentifier": "D:\\Dev\\html-webpack-plugin-manifest-issue\\index.js",
                "moduleName": "./index.js",
                "loc": "",
                "name": "main",
                "reasons": []
            }
        ]
    }
]
```

You can see that the parent has id `0`, however there is no chunk with id `0` hence the exception

**Comment by jantimon** [comment]
This seems to be a webpack or common chunks bug. we could ignore invalid parents here but I it would be way better to solve the root of the issue

**Comment by michael-wolfenden** [comment]
I'm not sure why this issue has been closed. The behavior of webpack has not changed between versions 1 and 2 (the invalid parent issue exists in both versions). The change to the default dependency sort is causing an issue such that what used to work for webpack 1 is broken for webpack 2.

**Comment by jantimon** [comment]
The sort order changed from webpack 1 to webpack 2 therefore we can't go with the old sorting either.

Please try to engage the root of the bug.

**Comment by michael-wolfenden** [comment]
@jantimon Apologies, I just got to the root of the issue and it has nothing to do with webpack.

If you look at the configuration, i'm excluding the `manifest` chunk. This is actually chunk `0`, the missing parent chunk.

The issue is that the plugin removes any excluded chunks **before** sorting, thus potentially removing parent chunks

I think the solution is to do the opposite, sort than exclude.

Thoughts?

**Comment by jantimon** [comment]
Oh nice that makes sense

**Comment by jantimon** [comment]
Didn't think about excludes - we could also update the parent child relations before sorting

**Comment by jantimon** [comment]
@michael-wolfenden would #343 solve this issue for you?

**Comment by michael-wolfenden** [comment]
@jantimon Yep, just tried it and it resolved the issue :ship:

**Comment by jantimon** [comment]
Cool 👍

**Comment by jantimon** [comment]
Shipped in 2.20
